transaction/src/manifest/generator.rs

```diff
@@ -1517,11 +1708,188 @@ where
     Ok(result)
 }
 
+pub fn generator_error_diagnostics(
+    s: &str,
+    err: GeneratorError,
+    style: CompileErrorDiagnosticsStyle,
+) -> String {
+    let (title, label) = match err.error_kind {
+        GeneratorErrorKind::InvalidAstType {
+            expected_type,
+            actual,
+        } => {
+            let title = format!(
+                "expected value kind {}, found {}",
+                expected_type.value_kind(),
+                actual.value_kind()
+            );
+            let label = format!(
+                "expected {}, found {}",
+                expected_type.value_kind(),
+                actual.value_kind()
+            );
+            (title, label)
+        }
+        GeneratorErrorKind::InvalidAstValue {
+            expected_type,
+            actual,
+        } => {
+            let mut types: Vec<String> = vec![];
+
+            // Remove potential duplicates.
+            // Duplicates might occure due to aliases,
+            // eg. PackageAddress, ComponentAddress are both Address
```

Comment:
Hopefully we won't need to do this any more when we remove `PackageAddress` and friends from `ast::ValueKind` which are relics of a historic incomplete refactor